include/fi_mem.h

```diff
@@ -69,24 +69,31 @@ static inline void *mem_dup(const void *src, size_t size)
  */
 #define FREESTACK_EMPTY	NULL
 
+#define FREESTACK_HEADER 					\
+	size_t		size;					\
+	void		*next;					\
+
 #define freestack_isempty(fs)	((fs)->next == FREESTACK_EMPTY)
 #define freestack_push(fs, p)					\
 {								\
 	*(void **) p = (fs)->next;				\
 	(fs)->next = p;						\
 }
-#define freestack_pop(fs)					\
-	({							\
-		void *p = (fs)->next;				\
-		assert(!freestack_isempty(fs));			\
-		(fs)->next = *((void **) p);			\
-		p;						\
-	})
+#define freestack_pop(fs) freestack_pop_impl(fs, fs->next)
+
+static inline void* freestack_pop_impl(void *fs, void *fs_next)
+{
+	struct {
+		FREESTACK_HEADER;
+	} *_fs = fs;
+	assert(!freestack_isempty(_fs));
+	(_fs)->next = *((void **)fs_next);
+	return fs_next;
+}
```

Comment:
I didn't take a look at the initial version of this patch but I have a question. Would it be a problem if we define freestack_pop(fs) directly as an inline function? Are we having the #define just to keep it as a macro to adhere to @sayantansur 's comment? 
nit: can we rename '_fs' to 'freestack'? I feel it would be make the code easier to read (not to confuse between fs_next and (_fs)->next)